Mathlib/Tactic/Push.lean

```diff
@@ -184,26 +185,6 @@ which will print the `push_neg` form of `e`.
 -/
 macro (name := pushNeg) tk:"#push_neg " e:term : command => `(command| #conv%$tk push_neg => $e)
 
-/-- Execute main loop of `push_neg` at the main goal. -/
-def pushNegTarget : TacticM Unit := withMainContext do
-  let goal ← getMainGoal
-  let tgt ← instantiateMVars (← goal.getType)
-  let newGoal ← applySimpResultToTarget goal tgt (← pushNegCore tgt)
-  if newGoal == goal then throwError "push_neg made no progress"
-  replaceMainGoal [newGoal]
-
-
-/-- Execute main loop of `push_neg` at a local hypothesis. -/
-def pushNegLocalDecl (fvarId : FVarId) : TacticM Unit := withMainContext do
-  let ldecl ← fvarId.getDecl
```

Comment:
Flagging for a final look: these three lines are handled differently in the new code.

Reply:
```
/-- error: push_neg made no progress -/
#guard_msgs in
def inductive_proof : ¬ False := by
  push_neg at inductive_proof
  trivial
```

The `push_neg` throws an error in this PR but not on current Mathlib. This seems like a desirable change, and more in line with the behaviour for other local declarations:

```
/-- error: push_neg made no progress -/
#guard_msgs in
def test : True := by
  have : True := trivial
  push_neg at this
  trivial
```

Reply:
Yep, I think it's a good idea to refuse to act on implementatation-detail hypotheses. I have now added this filtering, based on the version from `pushNegLocalDecl` but changing `isAuxDecl` to `isImplementationDetail` (which is more general):
```lean
  let ldecl ← fvarId.getDecl
  if ldecl.isImplementationDetail then
    throwError "cannot run {proc} at {ldecl.userName}, it is an implementation detail"
```
And I added Arend's test.

I don't think it matters that in the other use cases (e.g. `ring_nf`) this is a change in behaviour in edge cases. Part of the motivation for this PR is to allow us to implement best-practices across all similar tactic logic.